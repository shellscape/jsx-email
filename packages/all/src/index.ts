export * from '@jsx-email/body';
export * from '@jsx-email/button';
export * from '@jsx-email/column';
export * from '@jsx-email/container';
export * from '@jsx-email/font';
export * from '@jsx-email/head';
export * from '@jsx-email/heading';
export * from '@jsx-email/hr';
export * from '@jsx-email/html';
export * from '@jsx-email/img';
export * from '@jsx-email/link';
export * from '@jsx-email/markdown';
export * from '@jsx-email/preview';
export * from '@jsx-email/render';
export * from '@jsx-email/row';
export * from '@jsx-email/section';
export * from '@jsx-email/tailwind';
export * from '@jsx-email/text';
