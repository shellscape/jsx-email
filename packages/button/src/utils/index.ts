export * from './px-to-pt';
export * from './parse-padding';
