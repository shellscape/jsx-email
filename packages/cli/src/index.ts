#!/usr/bin/env node
import { parseArgs } from 'node:util';

import debugConfig from 'debug';
import importLocal from 'import-local';

import pkg from '../package.json';

import { command as build } from './commands/build';
import { command as create } from './commands/create';
import { command as help } from './commands/help';
import { command as preview } from './commands/preview';
import type { CommandFn } from './commands/types';

const commands: Record<string, CommandFn> = { build, create, help, preview };
const debug = debugConfig('@jsx-email/cli');
const { log } = console;

const run = async () => {
  const argv = parseArgs({ allowPositionals: true, args: process.argv.slice(2), strict: false });
  const [commandName] = argv.positionals;
  let command = commands[commandName];

  debug(`Command Name: \`${commandName}\``);

  if (argv.values.version) {
    log(`${pkg.name} v${pkg.version}\n`);
    return;
  }

  if (!command) command = help;

  const result = await command(argv.values, argv.positionals?.slice(1) || []);

  if (!result) {
    debug(`Command \`${commandName}\` returned \`false\``);
    help({}, []);
  }
};

if (importLocal(__filename)) {
  debug('Using local install of webpack-command');
} else {
  run();
}
