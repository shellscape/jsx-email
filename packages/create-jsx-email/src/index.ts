#!/usr/bin/env node
/* eslint-disable no-await-in-loop */
import { existsSync, readdirSync, rmSync } from 'fs';
import { mkdir, readFile, writeFile } from 'fs/promises';
import { basename, dirname, join, resolve, win32, posix } from 'path';

import chalk from 'chalk';
import { detect } from 'detect-package-manager';
import globby from 'globby';
import mustache from 'mustache';
import prompts from 'prompts';

import pkg from '../package.json';

interface CreateEmailArgs {
  jsx: boolean;
  name: string;
  outputPath: string;
}

const cancelled = () => {
  throw new Error(chalk`{red ✖} Operation cancelled`);
};
const defaultTargetDir = 'email-project';
const clearDirectory = (path: string) => {
  for (const file of readdirSync(path)) {
    // eslint-disable-next-line no-continue
    if (file === '.git') continue;
    rmSync(resolve(path, file), { force: true, recursive: true });
  }
};
const formatTargetDir = (targetPath?: string) => targetPath?.trim().replace(/\/+$/g, '');
const intro = chalk`
{underline create-jsx-email v${pkg.version}}

${pkg.description}
`;
const isEmpty = (path: string) => {
  const files = readdirSync(path);
  return files.length === 0 || (files.length === 1 && files[0] === '.git');
};
const { log } = console;
const normalizePath = (filename: string) => filename.split(win32.sep).join(posix.sep);
const typeDep = ',\n"@types/react": "^18.2.0",\n"typescript": "^5.2.2"';
const typeProps = `\nexport type TemplateProps = {
  email: string;
  name: string;
}`;

export const createEmail = async ({ jsx, name, outputPath }: CreateEmailArgs) => {
  const data = {
    name,
    propsType: jsx ? '' : ': TemplateProps',
    typeProps: jsx ? '' : typeProps
  };
  const templatePath = resolve(__dirname, '../generators/templates/email.mustache');
  const template = await readFile(templatePath, 'utf8');
  const contents = mustache.render(template, data);
  const fileName = basename(templatePath)
    .replace('_', '')
    .replace('.mustache', jsx ? '.jsx' : '.tsx');
  const outPath = join(outputPath, fileName);

  log('Creating a new template at', outPath);

  await writeFile(outPath, contents, 'utf8');

  return outPath;
};

const run = async () => {
  log(intro);

  const skip = process.argv.some((arg) => arg === '--yes');
  const argTargetDir = formatTargetDir(process.argv[2]);
  let targetPath = argTargetDir || defaultTargetDir;
  let result: prompts.Answers<'projectName' | 'projectType' | 'overwrite'>;
  const defaults: typeof result = {
    overwrite: true,
    projectName: defaultTargetDir,
    projectType: 'TypeScript'
  };

  try {
    result = skip
      ? defaults
      : await prompts(
          [
            {
              initial: defaultTargetDir,
              message: 'Project name:',
              name: 'projectName',
              onState: (state) => {
                targetPath = formatTargetDir(state.value) || defaultTargetDir;
              },
              type: argTargetDir ? null : 'text'
            },
            {
              choices: [
                { title: 'TypeScript', value: 'TypeScript' },
                { title: 'JavaScript', value: 'JavaScript' }
              ],
              message: 'TypeScript or JavaScript:',
              name: 'projectType',
              type: 'select'
            },
            {
              message: () =>
                targetPath === '.'
                  ? 'Current directory'
                  : `Target directory "${targetPath}" is not empty. Remove existing files and continue?`,
              name: 'overwrite',
              type: () => (!existsSync(targetPath) || isEmpty(targetPath) ? null : 'confirm')
            },
            {
              name: 'overwriteChecker',
              type: (_, { overwrite }: { overwrite?: boolean }) => {
                if (overwrite === false) cancelled();
                return null;
              }
            }
          ],
          { onCancel: cancelled }
        );
  } catch (cancelled: any) {
    log(cancelled.message);
    return;
  }

  const { overwrite, projectName, projectType } = result;
  const root = join(process.cwd(), targetPath);
  const generatorsPath = resolve(__dirname, '../generators');
  const jsx = projectType === 'JavaScript';
  const templates = await globby([normalizePath(join(generatorsPath, '/*.*'))]);
  const outputPath = join(root, 'templates');
  const templateData = { name: projectName, typeDep: jsx ? '' : typeDep };

  log(chalk`\n{blue Creating Project} at: {dim ${root}}`);

  if (overwrite && existsSync(root)) clearDirectory(root);

  await mkdir(outputPath, { recursive: true });

  for (const path of templates) {
    // eslint-disable-next-line no-continue
    if (path.endsWith('_tsconfig.json') && jsx) continue;
    const template = await readFile(path, 'utf8');
    const contents = mustache.render(template, templateData);
    const basePath = dirname(path);
    const fileName = basename(path).replace('_', '').replace('.mustache', '');
    const outPath = join(root, basePath.endsWith('templates') ? 'templates' : '', fileName);

    await writeFile(outPath, contents, 'utf8');
  }

  await createEmail({ jsx, name: projectName, outputPath });

  const packageManager = await detect();
  const install =
    packageManager === 'yarn'
      ? `  $ yarn\n  $ yarn dev`
      : `  $ ${packageManager} install\n  $ ${packageManager} run dev`;
  const footer = chalk`
{green ✓} {bold jsx-email} Project Created

Next, run:

  $ cd ${projectName}
${install}

{magenta Check out the docs!} {underline http://jsx.email/docs/quick-start}
`;

  log(footer);
};

run();
