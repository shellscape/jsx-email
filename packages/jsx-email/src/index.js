import './helpers.js';
// components
export * from './components/background.js';
export * from './components/body.js';
export * from './components/button.js';
export * from './components/butan.js';
export * from './components/code.js';
export * from './components/color-scheme.js';
export * from './components/column.js';
export * from './components/conditional.js';
export * from './components/container.js';
export * from './components/font.js';
export * from './components/graph.js';
export * from './components/head.js';
export * from './components/heading.js';
export * from './components/hr.js';
export * from './components/html.js';
export * from './components/img.js';
export * from './components/link.js';
export * from './components/markdown.js';
export * from './components/preview.js';
export * from './components/row.js';
export * from './components/raw.js';
export * from './components/section.js';
export * from './components/tailwind/tailwind.js';
export * from './components/text.js';
// renderer
export * from './renderer/compat/context.js';
export * from './renderer/compat/hooks.js';
export * from './renderer/jsx-to-string.js';
export * from './renderer/render.js';
export { useData } from './renderer/suspense.js';
export * from './plugins.js';
export * from './types.js';
//# sourceMappingURL=index.js.map