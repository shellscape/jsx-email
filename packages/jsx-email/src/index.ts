export * from './components/body';
export * from './components/button/button';
export * from './components/code';
export * from './components/color-scheme';
export * from './components/column';
export * from './components/conditional';
export * from './components/container';
export * from './components/font';
export * from './components/head';
export * from './components/heading';
export * from './components/hr';
export * from './components/html';
export * from './components/img';
export * from './components/link';
export * from './components/markdown';
export * from './components/preview';
export * from './components/row';
export * from './components/section';
export * from './components/tailwind/tailwind';
export * from './components/text';
export * from './render';
export * from './types';
