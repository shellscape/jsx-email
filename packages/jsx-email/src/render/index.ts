export * from './render';
export * from './jsx-to-string';
export * from './process';
