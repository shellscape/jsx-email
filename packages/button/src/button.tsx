import type React from 'react';

import { pxToPt, parsePadding } from './utils';

type RootProps = React.ComponentPropsWithoutRef<'a'>;

export interface ButtonProps extends RootProps {}

const buttonStyle = (
  style?: React.CSSProperties & {
    pb: number;
    pl: number;
    pr: number;
    pt: number;
  }
) => {
  const { pt, pr, pb, pl, padding, ...rest } = style || {};

  return {
    ...rest,
    display: 'inline-block',
    lineHeight: '100%',
    maxWidth: '100%',
    padding: `${pt}px ${pr}px ${pb}px ${pl}px`,
    textDecoration: 'none'
  };
};

const buttonTextStyle = (pb?: number) => {
  return {
    display: 'inline-block',
    lineHeight: '120%',
    maxWidth: '100%',
    msoPaddingAlt: '0px',
    msoTextRaise: pxToPt(pb || 0)
  };
};

export const Button: React.FC<Readonly<ButtonProps>> = ({
  children,
  style,
  target = '_blank',
  ...props
}) => {
  const { pt, pr, pb, pl } = parsePadding({
    padding: style?.padding,
    paddingBottom: style?.paddingBottom,
    paddingLeft: style?.paddingLeft,
    paddingRight: style?.paddingRight,
    paddingTop: style?.paddingTop
  });

  const y = pt + pb;
  const textRaise = pxToPt(y);

  return (
    <a
      {...props}
      data-id="@jsx-email/button"
      target={target}
      style={buttonStyle({ ...style, pb, pl, pr, pt })}
    >
      <span
        dangerouslySetInnerHTML={{
          __html: `<!--[if mso]><i style="letter-spacing: ${pl}px;mso-font-width:-100%;mso-text-raise:${textRaise}" hidden>&nbsp;</i><![endif]-->`
        }}
      />
      <span style={buttonTextStyle(pb)}>{children}</span>
      <span
        dangerouslySetInnerHTML={{
          __html: `<!--[if mso]><i style="letter-spacing: ${pr}px;mso-font-width:-100%" hidden>&nbsp;</i><![endif]-->`
        }}
      />
    </a>
  );
};

Button.displayName = 'Button';
