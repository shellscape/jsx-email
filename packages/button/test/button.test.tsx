import { render } from '@jsx-email/render';

import { Button } from '../src';

describe('<Button> component', () => {
  beforeEach(() => {
    vi.restoreAllMocks();
    vi.resetModules();
  });

  it('renders children correctly', () => {
    const testMessage = 'Test message';
    const html = render(<Button>{testMessage}</Button>);
    expect(html).toContain(testMessage);
  });

  it('passes style and other props correctly', () => {
    const style = { backgroundColor: 'red' };
    const html = render(
      <Button style={style} data-testid="button-test">
        Test
      </Button>
    );
    expect(html).toContain('background-color:red');
    expect(html).toContain('data-testid="button-test"');
  });

  it('renders correctly  with padding values from style prop', () => {
    const actualOutput = render(
      <Button style={{ padding: '12px 20px' }} href="https://example.com" />
    );
    expect(actualOutput).toMatchSnapshot();
  });

  it('renders with no padding value', () => {
    const actualOutput = render(<Button href="https://example.com" />);
    expect(actualOutput).toMatchSnapshot();
  });
});
