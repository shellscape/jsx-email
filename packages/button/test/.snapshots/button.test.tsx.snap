// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Button> component > renders correctly  with padding values from style prop 1`] = `"<!DOCTYPE html PUBLIC \\"-//W3C//DTD XHTML 1.0 Transitional//EN\\" \\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\\"><a href=\\"https://example.com\\" data-id=\\"@jsx-email/button\\" target=\\"_blank\\" style=\\"display:inline-block;line-height:100%;max-width:100%;padding:12px 20px 12px 20px;text-decoration:none\\"><span><!--[if mso]><i style=\\"letter-spacing: 20px;mso-font-width:-100%;mso-text-raise:18\\" hidden>&nbsp;</i><![endif]--></span><span style=\\"display:inline-block;line-height:120%;max-width:100%;mso-padding-alt:0px;mso-text-raise:9px\\"></span><span><!--[if mso]><i style=\\"letter-spacing: 20px;mso-font-width:-100%\\" hidden>&nbsp;</i><![endif]--></span></a>"`;

exports[`<Button> component > renders with no padding value 1`] = `"<!DOCTYPE html PUBLIC \\"-//W3C//DTD XHTML 1.0 Transitional//EN\\" \\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\\"><a href=\\"https://example.com\\" data-id=\\"@jsx-email/button\\" target=\\"_blank\\" style=\\"display:inline-block;line-height:100%;max-width:100%;padding:0px 0px 0px 0px;text-decoration:none\\"><span><!--[if mso]><i style=\\"letter-spacing: 0px;mso-font-width:-100%;mso-text-raise:0\\" hidden>&nbsp;</i><![endif]--></span><span style=\\"display:inline-block;line-height:120%;max-width:100%;mso-padding-alt:0px;mso-text-raise:0\\"></span><span><!--[if mso]><i style=\\"letter-spacing: 0px;mso-font-width:-100%\\" hidden>&nbsp;</i><![endif]--></span></a>"`;
